kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ActionsIterator.java

```diff
@@ -387,21 +388,56 @@ private CloseableIterator<ActionWrapper> getNextActionsIter() {
     }
   }
 
-  /** Take input (iterator<T>, boolean) and produce an iterator<T, boolean>. */
+  /**
+   * Take input (iterator<T>, boolean) and produce an iterator<T, boolean>. When the input iterator
+   * has been created from a delta file, this function will try to use the inCommitTimestamp (if
+   * available) from the CommitInfo action as the commit timestamp for the file.
+   */
   private CloseableIterator<ActionWrapper> combine(
       CloseableIterator<ColumnarBatch> fileReadDataIter,
       boolean isFromCheckpoint,
       long version,
       Optional<Long> timestamp) {
+    final List<ColumnarBatch> firstActionList = new ArrayList<>();
+    final int commitInfoOrdinal = deltaReadSchema.indexOf("commitInfo");
+    // For delta files, we want to use the inCommitTimestamp from commitInfo
+    // as the commit timestamp for the file.
+    // Since CommitInfo should be the first action in the delta when inCommitTimestamp is
+    // enabled, we will read the first action and try to extract the timestamp from it.
+    // We also cache the first action so that we can return it in the iterator.
+    Optional<Long> commitTimestamp = timestamp;
+    if (!isFromCheckpoint && fileReadDataIter.hasNext() && commitInfoOrdinal != -1) {
+      ColumnarBatch firstAction = fileReadDataIter.next();
+      firstActionList.add(firstAction);
+      ColumnVector commitInfoVector = firstAction.getColumnVector(commitInfoOrdinal);
+      Optional<Long> ictOpt =
+          InCommitTimestampUtils.getCommitInfo(commitInfoVector)
+              .flatMap(CommitInfo::getInCommitTimestamp);
+      if (ictOpt.isPresent()) {
+        commitTimestamp = ictOpt;
+      }
+    }
+    final Iterator<ColumnarBatch> firstActionIter = firstActionList.iterator();
```

Comment:
I think you can use:

`Utils.singletonCloseableIterator` to create a single-element closeable iterator

Then use `CloseableIterator.combine` to combine that iterator with `fileReadDataIter`

And then you just use that final iterator to fulfilll the `CloseableIterator<ActionWrapper>() {` implementation.

Seems simpler? no if statements?